apps/dcos_net/src/dcos_net_mesos_listener.erl

```diff
@@ -986,14 +1001,107 @@ stream(Data, #state{pid=Pid, size=Size, buf=Buf}=State) ->
     Buf0 = <<Buf/binary, Data/binary>>,
     case byte_size(Buf0) of
         BufSize when BufSize >= Size ->
-            <<Head:Size/binary, Tail/binary>> = Buf0,
-            State0 = State#state{size=undefined, buf=Tail},
-            try jiffy:decode(Head, [return_maps]) of Obj ->
-                {next, Obj, State0}
-            catch error:Error ->
-                {error, Error}
-            end;
+            stream_decode(Buf0, Size, State);
         _BufSize ->
             httpc:stream_next(Pid),
             {next, State#state{buf=Buf0}}
     end.
+
+-spec(stream_decode(binary(), integer(), State) -> {error, term()} |
+     {next, State} | {next, jiffy:object(), State}
+         when State :: state()).
+stream_decode(Buf, Size, State) ->
+    <<Head:Size/binary, Tail/binary>> = Buf,
+    State0 = State#state{size=undefined, buf=Tail},
+    try
+        prometheus_counter:inc(mesos_listener, messages_total, [], 1),
+        % is the Size right here?
+        prometheus_counter:inc(mesos_listener, bytes_total, [], Size)
+    catch error:_Error ->
+        ok
+    end,
+    try jiffy:decode(Head, [return_maps]) of Obj ->
+        {next, Obj, State0}
+    catch error:Error ->
+        {error, Error}
+    end.
+
+
+%%%===================================================================
+%%% Metrics functions
+%%%===================================================================
+
+-spec(init_metrics() -> ok).
+init_metrics() ->
```

Comment:
Had to break up this function due to size, but do not really like this approach. Do you have any alternatives?

Reply:
Could you please break it up by domain area, not by metric type?